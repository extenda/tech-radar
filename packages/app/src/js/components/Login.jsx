import { GoogleLogin } from '@react-oauth/google';
import { hostedDomain } from '@tech-radar/shared/src/auth-config';
import PropTypes from 'prop-types';
import React from 'react';

const Login = (props) => {
  const { onSuccess, onFailure } = props;
  return (
    <>
      <nav className="navbar" role="navigation">
        <ul className="breadcrumb">
          <li key="home">&nbsp;</li>
        </ul>
      </nav>
      <div className="container">
        <h1 className="center">Extenda Retail Tech Radar</h1>
        <div className="row">
          <div className="two columns">&nbsp;</div>
          <div className="five columns">
            <p>
              <span role="img" aria-label="Wave">
                &#x1F44B;
              </span>
              &nbsp; Welcome to the Extenda Retail Tech Radar. Login with your Extenda Retail Google
              account.
            </p>
          </div>
          <div className="three columns">
            <GoogleLogin
              className="google-button"
              onSuccess={onSuccess}
              onFailure={onFailure}
              theme="outline"
              hosted_domain={hostedDomain}
            />
          </div>
          <div className="two columns">&nbsp;</div>
        </div>
      </div>
    </>
  );
};

Login.propTypes = {
  onSuccess: PropTypes.func.isRequired,
  onFailure: PropTypes.func.isRequired,
};

export default Login;
