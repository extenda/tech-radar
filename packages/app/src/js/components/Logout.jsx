import { googleLogout } from '@react-oauth/google';
import React from 'react';

import Icon from './Icon';

const logout = () => {
  googleLogout();
  window.setTimeout(() => {
    window.location.reload();
  }, 100);
};

const Logout = () => (
  <button type="button" className="google-button signout" onClick={logout}>
    <Icon name="sign-out" />
    Sign out
  </button>
);

export default Logout;
