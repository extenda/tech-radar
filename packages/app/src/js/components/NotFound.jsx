import React from 'react';

import radarService from '../modules/radarService';
import Navigation from './Navigation';

const NotFound = () => {
  const radar = radarService.model;

  return (
    <>
      <Navigation home={radar.quadrants} />
      <div className="container">
        <div className="row">
          <div className="container">
            <h1>Not Found</h1>
            <p>
              The requested resource was not found. You might have mistyped
              something or found a broken link.
            </p>
          </div>
        </div>
      </div>
    </>
  );
};

export default NotFound;
