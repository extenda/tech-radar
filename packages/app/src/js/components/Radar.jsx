import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withLDConsumer } from 'launchdarkly-react-client-sdk';
import { pick } from '@tech-radar/shared/src/utils';
import SvgRadar from '../lib/radar';
import radarService from '../modules/radarService';
import Icon from './Icon';
import Navigation from './Navigation';
import TagsInput from './TagsInput';
import QuadrantList from './QuadrantList';

export class Radar extends Component {
  constructor(props) {
    super(props);
    this.svgRadar = this.createSvgRadar();
    this.state = {
      blips: radarService.listBlips([]),
      tags: radarService.listTags(),
    };
  }

  createSvgRadar = () => {
    const { history } = this.props;

    const radar = radarService.model;

    return new SvgRadar({
      svg_id: 'radar',
      width: 900, // 1450 Set to 900 if not printing labels.
      height: 900,
      colors: {
        background: '#fff',
        grid: '#bbb',
        inactive: '#ddd',
      },
      quadrants: radar.quadrants.map((q) => pick(q, 'name')),
      rings: [
        { name: 'Adopt', color: '#43a047' },
        { name: 'Trial', color: '#c0ca33' },
        { name: 'Assess', color: '#ffb300' },
        { name: 'Hold', color: '#bf360c' },
      ],
      print_layout: true,
      print_entities: false,
      print_footer: true,
      linkOnClick: (link) => {
        history.push(link);
      },
    });
  };

  renderRadar = () => {
    this.svgRadar.renderBackground();
    this.renderBlips();
  };

  renderBlips = () => {
    const { blips } = this.state;
    this.svgRadar.renderEntries(blips);
  };

  onFilter = (tags) => {
    this.setState((prevState) => ({
      ...prevState,
      blips: radarService.listBlips(tags.map((t) => t.name)),
    }), this.renderBlips);
  };

  onChangeRadar = (e) => {
    // We must trigger a change on the TagFilter component to update the blips!
    const id = e.currentTarget.getAttribute('data-radar-id');
    if (radarService.model.id !== id) {
      radarService.switchRadar(id).then(() => {
        this.setState((prevState) => ({
          ...prevState,
          tags: radarService.listTags(),
          blips: radarService.listBlips(),
        }), this.renderBlips);
      });
    }
  }

  componentDidMount = () => {
    this.renderRadar();
  };

  onMouseEnter = (e) => {
    const id = e.currentTarget.getAttribute('data-radar-id');
    this.svgRadar.showBubbleForId(id);
    this.svgRadar.highlightLegendItem({ id });
  };

  onMouseLeave = (e) => {
    const id = e.currentTarget.getAttribute('data-radar-id');
    this.svgRadar.hideBubble();
    this.svgRadar.unhighlightLegendItem({ id });
  };

  renderQuadrant = (quadrant) => {
    const { blips } = this.state;
    const entries = radarService.getQuadrant(quadrant.dirname, true);

    return (
      <>
        <h2>{quadrant.name}</h2>
        <QuadrantList
          quadrant={entries}
          blips={blips}
          headerLevel={3}
          useShortname
          onMouseEnter={this.onMouseEnter}
          onMouseLeave={this.onMouseLeave}
        />
      </>
    );
  }

  render = () => {
    const radar = radarService.model;

    const { flags } = this.props;
    const { tags } = this.state;

    return (
      <>
        <Navigation home={radar.quadrantsNavBar} />
        {flags.releaseToolRadar && (
          <>
            <div className="u-full-width" style={{ marginBottom: '20px' }}>
              <center>
                <button type="button" data-radar-id="radar" onClick={this.onChangeRadar}>
                  <Icon name="rss" />
                  Engineering
                </button>
                <button type="button" data-radar-id="radar_it" onClick={this.onChangeRadar}>
                  <Icon name="signal" />
                  IT & BIS
                </button>
              </center>
            </div>
            <h1 className="center">
              {radar.title}
              &nbsp;
              <span className="version">
                <Icon name="calendar" />
                {radar.formattedDate}
              </span>
            </h1>
          </>
        )}
        {!flags.releaseToolRadar && (
        <h1 className="center">
          {radar.title}
          &nbsp;
          <span className="version">
            <Icon name="tag" />
            {radar.version}
          </span>
        </h1>
        )}
        <div className="tags-filter">
          <TagsInput onFilter={this.onFilter} tags={tags} />
        </div>
        <div className="radar-container u-full-width">
          <div className="radar-entries">
            {this.renderQuadrant(radar.quadrants[2])}
            {this.renderQuadrant(radar.quadrants[1])}
          </div>
          <div className="svg-radar">
            <svg id="radar" />
          </div>
          <div className="radar-entries">
            {this.renderQuadrant(radar.quadrants[3])}
            {this.renderQuadrant(radar.quadrants[0])}
          </div>
        </div>
        <div className="u-cf" />
        <div className="container">
          <div className="row">
            <div className="one-half column">
              <h2>About the Tech Radar</h2>
              <p>
                The Extenda Retail Tech Radar is based on the concept introduced by
                {' '}
                <a href="https://www.thoughtworks.com/radar">ThoughtWorks</a>
                .
                The Radar contains a list of technologies, complemented by an assessment result,
                called ring assessment.
                The following rings are used:
              </p>
              <ul className="no-bullets">
                <li>
                  <strong>Adopt</strong>
                  {' — '}
                  Technologies we have high confidence in and recommend for wide use.
                </li>
                <li>
                  <strong>Trial</strong>
                  {' — '}
                  Technologies we have seen used with success, but are yet to be proved successfully
                  in wide use and at large scale. Carries higher risk.
                </li>
                <li>
                  <strong>Assess</strong>
                  {' — '}
                  Technologies that are promising with a clear potential value for us.
                  Requires research and prototyping to analyze its impact.
                </li>
                <li>
                  <strong>Hold</strong>
                  {' — '}
                  Technologies not recommended for new products or technology we no longer
                  want to invest in. Existing products may still consider these depending
                  on their legacy.
                </li>
              </ul>
            </div>
            <div className="one-half column">
              <h2>The Purpose</h2>
              <p>
                The Tech Radar aims to inspire and support teams at Extenda Retail to pick the
                best technologies for their projects. It is way to share experience and knowledge
                between the teams and to create transparency about the technology direction of
                Extenda Retail.
                The Tech Radar also becomes a list of dos and don&apos;ts, what to try and to avoid
                in order to increase success.
              </p>
              <h2>How to Contribute</h2>
              <p>
                The Tech Radar is maintained on
                {' '}
                <a href="https://github.com/extenda/tech-radar">GitHub</a>
                .
                There you can read more about the process and how to contribute.
              </p>
            </div>
          </div>
        </div>
      </>
    );
  };
}

Radar.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  flags: PropTypes.object,
};

Radar.defaultProps = {
  flags: {},
};

export default withLDConsumer()(Radar);
