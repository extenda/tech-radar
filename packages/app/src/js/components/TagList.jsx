import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';

import radarService from '../modules/radarService';
import Icon from './Icon';
import Navigation from './Navigation';

const TagList = (props) => {
  const {
    match: {
      params: { tag },
    },
  } = props;

  const entries = radarService.listEntriesByTag(tag);

  return (
    <>
      <Navigation tag={tag} />
      <div className="container">
        <h1>
          <Icon name="tag" />
          {tag}
        </h1>
        <ul>
          {entries.map((entry) => (
            <li key={entry.name}>
              <Link to={`/entries/${entry.filename}`}>{entry.name}</Link>
            </li>
          ))}
        </ul>
      </div>
    </>
  );
};

TagList.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      tag: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

export default TagList;
