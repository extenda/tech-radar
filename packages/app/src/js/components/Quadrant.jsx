import PropTypes from 'prop-types';
import React from 'react';

import radarService from '../modules/radarService';
import Navigation from './Navigation';
import NotFound from './NotFound';
import QuadrantList from './QuadrantList';

const Quadrant = ({ match }) => {
  const quadrant = radarService.getQuadrant(match.params.quadrant);

  if (!quadrant) {
    return <NotFound />;
  }

  return (
    <>
      <Navigation quadrant={{ name: quadrant.name, dirname: quadrant.dirname }} />
      <div className="container">
        <h1>{quadrant.name}</h1>
        <QuadrantList quadrant={quadrant} />
      </div>
    </>
  );
};

Quadrant.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      quadrant: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

export default Quadrant;
