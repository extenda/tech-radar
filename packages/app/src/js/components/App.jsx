import { jwtDecode } from 'jwt-decode';
import { withLDConsumer } from 'launchdarkly-react-client-sdk';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';

import radarService from '../modules/radarService';
import sha256 from '../modules/sha256';
import Entry from './Entry';
import Footer from './Footer';
import Login from './Login';
import Logout from './Logout';
import NotFound from './NotFound';
import Quadrant from './Quadrant';
import LDRadar from './Radar';
import TagList from './TagList';

export class App extends Component {
  constructor(props) {
    super(props);
    if (typeof DISABLE_GOOGLE_LOGIN === 'boolean' && DISABLE_GOOGLE_LOGIN) {
      // No log in required. Usually desired in development server.
      this.state = {
        isSignedIn: true,
        loading: true,
      };
      radarService.init().then(this.radarDidLoad);
    } else {
      this.state = {
        isSignedIn: false,
        loading: true,
      };
    }
  }

  radarDidLoad = () => {
    this.setState((prevState) => ({
      ...prevState,
      loading: false,
    }));
  };

  radarDidFail = (err) => {
    // eslint-disable-next-line no-console
    console.error('Failed to load radar entries', err);
    this.setState((prevState) => ({
      ...prevState,
      isSignedIn: false,
      loading: true,
    }));
  };

  loginDidSucceed = (response) => {
    this.setState((prevState) => ({
      ...prevState,
      isSignedIn: true,
    }));

    const { ldClient } = this.props;
    if (!ldClient) {
      // eslint-disable-next-line no-console
      console.debug('Missing LDClient property');
      return null;
    }

    const { sub, email } = jwtDecode(response.credential);
    return sha256(sub).then((key) =>
      ldClient
        .identify({
          kind: 'user',
          key,
          email,
          _meta: {
            privateAttributeNames: ['email'],
          },
        })
        .then(() =>
          radarService.init(response.credential).then(this.radarDidLoad).catch(this.radarDidFail),
        ),
    );
  };

  loginDidFail = () => {};

  render = () => {
    const { loading, isSignedIn } = this.state;

    if (!isSignedIn) {
      return <Login onSuccess={this.loginDidSucceed} onFailure={this.loginDidFail} />;
    }

    if (loading) {
      return null;
    }

    return (
      <>
        <Logout />
        <BrowserRouter>
          <Switch>
            <Route exact path="/" component={LDRadar} />
            <Route path="/entries/:id" component={Entry} />
            <Route path="/:quadrant.html" component={Quadrant} />
            <Route path="/tags/:tag.html" component={TagList} />
            <Route path="*" component={NotFound} />
          </Switch>
        </BrowserRouter>
        <Footer />
      </>
    );
  };
}

App.propTypes = {
  ldClient: PropTypes.shape({
    identify: PropTypes.func.isRequired,
  }),
};
App.defaultProps = {
  ldClient: null,
};

export default withLDConsumer({ clientOnly: true })(App);
