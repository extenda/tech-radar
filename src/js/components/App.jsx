import React, { Component } from 'react';
import {
  BrowserRouter,
  NavLink,
  Switch,
  Route,
} from 'react-router-dom';
import Radar from './Radar';
import Entry from './Entry';
import Quadrant from './Quadrant';
import Footer from './Footer';
import NotFound from './NotFound';
import radarService from '../modules/radarService';

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
    };
  }

  radarDidLoad = () => {
    this.setState({
      loading: false,
    });
  };

  componentDidMount = () => radarService.init().then(this.radarDidLoad);

  render = () => {
    const { loading } = this.state;

    if (loading) {
      return null;
    }

    return (
      <React.Fragment>
        <BrowserRouter>
          <Switch>
            <Route exact path="/" component={Radar} />
            <Route path="/entries/:id" component={Entry} />
            <Route path="/:quadrant.html" component={Quadrant} />
            <Route path="*" component={NotFound} />
          </Switch>
        </BrowserRouter>
        <Footer />
      </React.Fragment>
    );
  };
}
