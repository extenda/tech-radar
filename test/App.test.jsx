import React from 'react';
import * as reactRouter from 'react-router-dom';
import { shallow, mount } from 'enzyme';
import App from '../src/js/components/App';
import Radar from '../src/js/components/Radar';
import Entry from '../src/js/components/Entry';
import Quadrant from '../src/js/components/Quadrant';
import NotFound from '../src/js/components/NotFound';
import TagList from '../src/js/components/TagList';
import Logout from '../src/js/components/Logout';
import Login from '../src/js/components/Login';

const { MemoryRouter } = reactRouter;

// eslint-disable-next-line react/prop-types
const mockRouter = initialEntry => ({ children }) => (
  <MemoryRouter initialEntries={[initialEntry]}>
    {children}
  </MemoryRouter>
);

jest.mock('../src/js/components/Radar', () => () => (
  <div>
    Radar
  </div>
));

jest.mock('../src/js/components/Entry', () => () => (
  <div>
    Entry
  </div>
));

jest.mock('../src/js/components/Quadrant', () => () => (
  <div>
    Quadrant
  </div>
));

jest.mock('../src/js/components/NotFound', () => () => (
  <div>
    NotFound
  </div>
));

jest.mock('../src/js/components/TagList', () => () => (
  <div>
    TagList
  </div>
));

jest.mock('../src/js/components/Login', () => () => (
  <div>
    Login
  </div>
));

jest.mock('../src/js/components/Logout', () => () => (
  <div>
    Logout
  </div>
));

jest.mock('../src/js/modules/radarService');

describe('<App />', () => {
  let spyConsole;
  beforeAll(() => {
    spyConsole = jest.spyOn(console, 'error').mockImplementation();
  });
  afterAll(() => {
    spyConsole.mockRestore();
  });

  test('It renders <Login /> if not authenticated', async () => {
    const component = mount(<App />);
    expect(component.state().isSignedIn).toEqual(false);
    expect(component).toMatchSnapshot();
  });

  test('It renders nothing before radar is loaded', async () => {
    const component = shallow(<App />);
    component.setState({ isSignedIn: true, loading: true });
    await component.update();
    expect(component.state().isSignedIn).toEqual(true);
    expect(component.state().loading).toEqual(true);
    expect(component).toMatchSnapshot();
  });

  test('It renders Radar on successful login', async () => {
    reactRouter.BrowserRouter = mockRouter('/');
    const component = mount(<App />);
    await component.instance().loginDidSucceed({ tokenId: 'test' });
    await component.update();
    expect(component.state().isSignedIn).toEqual(true);
    expect(component.state().loading).toEqual(false);
    expect(component.find(Logout)).toHaveLength(1);
    expect(component.find(Radar)).toHaveLength(1);
  });

  test('It does not render Radar on invalid JWT', async () => {
    reactRouter.BrowserRouter = mockRouter('/');
    const component = mount(<App />);
    await component.instance().loginDidSucceed({ tokenId: 'fail' });
    await component.update();
    expect(component.state().isSignedIn).toEqual(false);
    expect(component.state().loading).toEqual(true);

    // We need to render one more time to update for the state set from the failure.
    await component.update();
    expect(component.find(Login)).toHaveLength(1);
  });

  test('It renders the radar on / route', async () => {
    reactRouter.BrowserRouter = mockRouter('/');
    const component = mount(<App />);
    component.setState({ isSignedIn: true, loading: false });
    await component.update();
    expect(component.state().isSignedIn).toEqual(true);
    expect(component.find(Radar)).toHaveLength(1);
  });

  test('It renders <Logout /> on / route', async () => {
    reactRouter.BrowserRouter = mockRouter('/');
    const component = mount(<App />);
    component.setState({ isSignedIn: true, loading: false });
    await component.update();
    expect(component.state().isSignedIn).toEqual(true);
    expect(component.find(Logout)).toHaveLength(1);
  });

  test('It renders quadrant on /*.html route', async () => {
    reactRouter.BrowserRouter = mockRouter('/dev.html');
    const component = mount(<App />);
    component.setState({ isSignedIn: true, loading: false });
    expect(component.find(Quadrant)).toHaveLength(1);
  });

  test('It renders entry on /entries/*.html route', async () => {
    reactRouter.BrowserRouter = mockRouter('/entries/java.html');
    const component = mount(<App />);
    component.setState({ isSignedIn: true, loading: false });
    await component.update();
    expect(component.find(Entry)).toHaveLength(1);
  });

  test('It renders not found for invalid URL', async () => {
    reactRouter.BrowserRouter = mockRouter('/notfound');
    const component = mount(<App />);
    component.setState({ isSignedIn: true, loading: false });
    await component.update();
    expect(component.find(NotFound)).toHaveLength(1);
  });

  test('It renders tag list no /tags/*.html route', async () => {
    reactRouter.BrowserRouter = mockRouter('/tags/java.html');
    const component = mount(<App />);
    component.setState({ isSignedIn: true, loading: false });
    await component.update();
    expect(component.find(TagList)).toHaveLength(1);
  });
});
