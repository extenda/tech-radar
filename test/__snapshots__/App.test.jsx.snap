// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> It renders nothing before radar is loaded 1`] = `""`;
